### PR TITLE
2.6 - Watch Unit Config Changes Via Model Cache

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch changes the uniter API server `WatchConfigSettingsHash` to watch the model cache instead of state.

## QA steps

- Bootstrap
- `juju deploy redis`.
- Run `juju debug-log --include redis/0` in a separate terminal and observe that the config-changed hook fires for each of the configuration change actions below.
- `juju config redis password=new-pass`.
- `juju ssh 0 cat /etc/redis/redis-charm.conf` and check that the line "requirepass new-pass" is present.
- `juju config redis databases=4`.
- `juju ssh 0 cat /etc/redis/redis-charm.conf` and check that the line "databases 4" is present.
- `export JUJU_DEV_FEATURE_FLAGS=generations`.
- `juju add-branch test-branch`.
- `juju track test-branch redis/0`.
- `juju config redis password=branch-pass --branch test-branch`.
- `juju ssh 0 cat /etc/redis/redis-charm.conf` and check that the line "requirepass branch-pass" is present.
- `juju config redis databases=8 --branch test-branch`.
- `juju ssh 0 cat /etc/redis/redis-charm.conf` and check that the line "databases 8" is present;

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Replace State Unit Config Watcher with Cache Implementation](https://trello.com/c/myis3fkm/1549-replace-state-unit-config-watcher-with-cache-implementation)

**Comment by mitechie** [comment]
!!build!!

**Comment by mitechie** [comment]
QA
---
With the original instructions on the redis charm I got a bunch of errors:

    unit-redis-1: 15:03:56 INFO juju.worker.uniter.charm downloading cs:trusty/redis-0 from API server
    unit-redis-1: 15:03:56 INFO juju.downloader downloading from cs:trusty/redis-0
    unit-redis-1: 15:03:56 INFO juju.downloader download complete ("cs:trusty/redis-0")
    unit-redis-1: 15:03:56 INFO juju.downloader download verified ("cs:trusty/redis-0")
    unit-redis-1: 15:03:56 INFO juju.worker.uniter hooks are retried true
    machine-1: 15:04:00 ERROR juju.worker.dependency "broker-tracker" manifold worker returned unexpected error: cannot create instance broker: creating LXD container manager: Get http://unix.socket/1.0: dial unix /var/lib/lxd/unix.socket: connect: no such file or directory
    machine-1: 15:04:05 ERROR juju.worker.dependency "broker-tracker" manifold worker returned unexpected error: cannot create instance broker: creating LXD container manager: Get http://unix.socket/1.0: dial unix /var/lib/lxd/unix.socket: connect: no such file or directory
    machine-0: 15:04:05 ERROR juju.worker.dependency "broker-tracker" manifold worker returned unexpected error: cannot create instance broker: creating LXD container manager: Get http://unix.socket/1.0: dial unix /var/lib/lxd/unix.socket: connect: no such file or directory
    machine-1: 15:04:11 ERROR juju.worker.dependency "broker-tracker" manifold worker returned unexpected error: cannot create instance broker: creating LXD container manager: Get http://unix.socket/1.0: dial unix /var/lib/lxd/unix.socket: connect: no such file or directory
    machine-1: 15:04:19 ERROR juju.worker.dependency "broker-tracker" manifold worker returned unexpected error: cannot create instance broker: creating LXD container manager: Get http://unix.socket/1.0: dial unix /var/lib/lxd/unix.socket: connect: no such file or directory
    machine-0: 15:04:27 ERROR juju.worker.dependency "broker-tracker" manifold worker returned unexpected error: cannot create instance broker: creating LXD container manager: Get http://unix.socket/1.0: dial unix /var/lib/lxd/unix.socket: connect: no such file or directory

Trying out another charm to avoid the trusty/lxd issue I might be hitting on my GCE test environment.

**Comment by mitechie** [comment]
!!build!! failed to start mongo...

**Comment by mitechie** [comment]
QA take 2 - grafana

Made an initial admin_password config change and saw both units update

    unit-grafana-0: 15:43:07 INFO unit.grafana/0.juju-log [*] admin password updated on database
    unit-grafana-1: 15:43:07 INFO unit.grafana/1.juju-log [*] admin password updated on database

Created the branch, tracked unit 1, and then changed the password a second time. 

    unit-grafana-1: 15:45:03 INFO unit.grafana/1.juju-log [*] admin password updated on database

checked diff output

    % juju diff
    change-pass:
      created: 25 Jun 2019 15:44:07-04:00
      created-by: admin
      applications:
      - application: grafana
        progress: 1/2
        config:
          admin_password: steptwo

Then ran Juju commit on the branch to see the change go out to unit 0

    unit-grafana-0: 15:48:03 INFO unit.grafana/0.juju-log [*] admin password updated on database

QA 👍

File: api/uniter/unit_test.go [hunk 1]
```diff
@@ -547,24 +547,36 @@ func (s *unitSuite) TestWatchConfigSettingsHash(c *gc.C) {
...
 	err = s.wordpressApplication.UpdateCharmConfig(model.GenerationMaster, charm.Settings{
 		"blog-title": "sauceror central",
 	})
+
+	s.State.StartSync()
+	s.WaitForModelWatchersIdle(c, s.Model.UUID())
+
 	c.Assert(err, jc.ErrorIsNil)
-	wc.AssertChange("7ed6151e9c3d5144faf0946d20c283c466b4885dded6a6122ff3fdac7ee2334f")
+	wc.AssertChange("a5ac26f48388f7e22ce2c08e94af6ca92b128467542977d3f6d18cd9121cb0f4")
```
Comment (howbazaar) [inline api/uniter/unit_test.go]: I had a feeling we'd have different hash values. Is there a way they could be brought in line? I feel I have to ask.

File: core/cache/stringswatcher_test.go [hunk 2]
```diff
@@ -1,113 +0,0 @@
-// Copyright 2018 Canonical Ltd.
-// Licensed under the AGPLv3, see LICENCE file for details.
-
-package cache_test
```
Comment (howbazaar) [inline core/cache/stringswatcher_test.go]: Why not just change the package rather than moving the contents?

Comment (manadart) [reply core/cache/stringswatcher_test.go]: Yes, I could have done that.

File: api/uniter/unit_test.go (hunk 1 above)
Comment (manadart) [reply api/uniter/unit_test.go]: The "old" implementation is to convert the settings map to a sorted slice of `bson.DocElem`, then marshal it to a BSON byte slice, versus marshalling a map to a YAML byte slice. Both then get a SHA256 hash of the result.

I guess we could keep the implementations consistent - probably a negligible overhead if any, but if feels a bit icky bringing such a dependency into core.

I will land this and we can change it if easily if desired.

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
`fatal: unable to access 'https://github.com/juju/juju.git/': Could not resolve host: github.com`
$$merge$$

**Pull request merged**